From 93b5b6893c7c057dcfa4d09d9ea3a5d0e3ba9965 Mon Sep 17 00:00:00 2001
From: Steve Dickson <SteveD@redhat.com>
Date: Thu, 18 Sep 2014 14:19:40 -0400
Subject: [fs] pnfs/filelayout: retry ds commit if nfs_commitdata_alloc fails

Message-id: <1411050057-12417-18-git-send-email-steved@redhat.com>
Patchwork-id: 96192
O-Subject: [RHEL7.1 PATCH 17/94] pnfs/filelayout: retry ds commit if nfs_commitdata_alloc fails
Bugzilla: 1111708
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>

From: Peng Tao <tao.peng@primarydata.com>

Signed-off-by: Peng Tao <tao.peng@primarydata.com>
Signed-off-by: Tom Haynes <Thomas.Haynes@primarydata.com>
Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Upstream Id: 0b0bc6ea77ec8430626e8c2f6cbdfb767b168ac1
Bugzilla: 1111708
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfs/filelayout/filelayout.c b/fs/nfs/filelayout/filelayout.c
index a928f92..2576d28 100644
--- a/fs/nfs/filelayout/filelayout.c
+++ b/fs/nfs/filelayout/filelayout.c
@@ -1237,15 +1237,33 @@ restart:
  spin_unlock(cinfo->lock);
 }
 
+static void filelayout_retry_commit(struct nfs_commit_info *cinfo, int idx)
+{
+ struct pnfs_ds_commit_info *fl_cinfo = cinfo->ds;
+ struct pnfs_commit_bucket *bucket = fl_cinfo->buckets;
+ struct pnfs_layout_segment *freeme;
+ int i;
+
+ for (i = idx; i < fl_cinfo->nbuckets; i++, bucket++) {
+  if (list_empty(&bucket->committing))
+   continue;
+  nfs_retry_commit(&bucket->committing, bucket->clseg, cinfo);
+  spin_lock(cinfo->lock);
+  freeme = bucket->clseg;
+  bucket->clseg = NULL;
+  spin_unlock(cinfo->lock);
+  pnfs_put_lseg(freeme);
+ }
+}
+
 static unsigned int
 alloc_ds_commits(struct nfs_commit_info *cinfo, struct list_head *list)
 {
  struct pnfs_ds_commit_info *fl_cinfo;
  struct pnfs_commit_bucket *bucket;
  struct nfs_commit_data *data;
- int i, j;
+ int i;
  unsigned int nreq = 0;
- struct pnfs_layout_segment *freeme;
 
  fl_cinfo = cinfo->ds;
  bucket = fl_cinfo->buckets;
@@ -1265,16 +1283,7 @@ alloc_ds_commits(struct nfs_commit_info *cinfo, struct list_head *list)
  }
 
  /* Clean up on error */
- for (j = i; j < fl_cinfo->nbuckets; j++, bucket++) {
-  if (list_empty(&bucket->committing))
-   continue;
-  nfs_retry_commit(&bucket->committing, bucket->clseg, cinfo);
-  spin_lock(cinfo->lock);
-  freeme = bucket->clseg;
-  bucket->clseg = NULL;
-  spin_unlock(cinfo->lock);
-  pnfs_put_lseg(freeme);
- }
+ filelayout_retry_commit(cinfo, i);
  /* Caller will clean up entries put on list */
  return nreq;
 }
@@ -1294,8 +1303,12 @@ filelayout_commit_pagelist(struct inode *inode, struct list_head *mds_pages,
    data->lseg = NULL;
    list_add(&data->pages, &list);
    nreq++;
-  } else
+  } else {
    nfs_retry_commit(mds_pages, NULL, cinfo);
+   filelayout_retry_commit(cinfo, 0);
+   cinfo->completion_ops->error_cleanup(NFS_I(inode));
+   return -ENOMEM;
+  }
  }
 
  nreq += alloc_ds_commits(cinfo, &list);
-- 
1.7.1